src/cargo/util/toml/mod.rs

```diff
@@ -94,11 +105,71 @@ fn read_manifest_from_str(
 
     let mut unused = BTreeSet::new();
     let deserializer = toml::de::Deserializer::new(contents);
-    let manifest: manifest::TomlManifest = serde_ignored::deserialize(deserializer, |path| {
+    let manifest: manifest::TomlManifest = match serde_ignored::deserialize(deserializer, |path| {
         let mut key = String::new();
         stringify(&mut key, &path);
         unused.insert(key);
-    })?;
+    }) {
+        Ok(manifest) => manifest,
+        Err(e) => {
+            let Some(span) = e.span() else {
+                return Err(e.into());
+            };
+
+            let (line_num, column) = translate_position(contents, span.start);
+            let num_new_lines = contents[span.start..span.end]
+                .chars()
+                .filter(|c| c == &'\n')
+                .count()
+                - (contents[span.end - 1..span.end].ends_with('\n') as usize);
+            let lines = contents.lines();
+            let source: String = lines
+                .skip(line_num)
+                .take(num_new_lines + 1)
+                .enumerate()
+                .map(|(i, line)| {
+                    let mut line = line.to_string();
+                    // Add back the new line that was stripped by `.lines()
+                    if num_new_lines > i {
+                        line.push('\n');
+                    }
+                    line
+                })
+                .collect();
+            // Make sure we don't try to highlight past the end of the line,
+            // but also make sure we are highlighting at least one character
+            let highlight_end = (column + (span.end - span.start))
```

Comment:
bug: `column` was calculated with `chars()` but the `span` math is using bytes

Reply:
imo `annotate-snippets` should be taking a byte span

Reply:
`rustc` uses char positions when rendering their diagnostics. They have a method of getting a `CharPos` from a byte. For consistency, we should probably do the same.

Reply:
From an API design perspective, are we intending the caller to describe the visual range to be highlighted of the slice within the content that should be highlighted?

While this might require a breaking change, imo the caller shouldn't be concerned with the exact rendering policy like that and it should be left up to `annotate-snippets`.

Reply:
I will open an issue for this in `annotate-snippets` to switch to bytes long-term. In the meantime, it shouldn't cause too many problems within `cargo`.

Reply:
My immediate concern with this code is that we are mixing units: `column` is `char` based but the span math is `u8` based.

Whatever way we go long term, we should at least pass in the data for the units `annotate-snippets` wants today and not mixed units.

Reply:
The most recent change makes it so we are not mixing units